### PR TITLE
Support batching for `RANGE` running window aggregations. Including on [databricks]

### PR DESCRIPTION
**Pull request by mythrocks**
This is a followup to #9489, but applies to more than `FIRST()`.

`GpuRunningWindowExec` handles "running-window" aggregations, i.e. `[UNBOUNDED PRECEDING, CURRENT ROW`, as long as they are `ROW` based window specifications. 

But when the window spec isn't specified, the default is to use `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`:
```sql
SELECT FIRST(foo) OVER (PARTITION BY bar ORDER BY goo) ...
-- Defaults to RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW.
```

This causes the window aggregation to be done with a single batch, with an increased likelihood of OOMs.

This commit adds batching support for `RANGE` queries by including the order-by column in batching. This allows the rows with the same values of partition-keys and order-by-keys to be processed as part of the same batch.  This allows us to avoid crossing batches for fix-ups.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala [hunk 1]
```diff
@@ -1668,6 +1675,28 @@ case class GpuRunningWindowExec(
 
   override def otherCopyArgs: Seq[AnyRef] = cpuPartitionSpec :: cpuOrderSpec :: Nil
 
+  override def childrenCoalesceGoal: Seq[CoalesceGoal] = Seq(outputBatching)
+
+  override def outputBatching: CoalesceGoal = {
+    val isRangeFrame = windowOps.exists {
+      case GpuAlias(
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:1682]: Are we guaranteed that this is always the pattern? If we miss a range based window, then we would end up with data corruption, so I just want to be sure that we are being very robust.

**Review by revans2 — COMMENTED** [review]
Looking good.

**Comment by mythrocks** [comment]
Build

File: integration_tests/src/main/python/window_function_test.py [hunk 2]
```diff
@@ -514,6 +515,37 @@ def test_window_running_no_part(b_gen, batch_size):
...
+
+    if isinstance(b_gen.data_type, NumericType) and not isinstance(b_gen, FloatGen) and not isinstance(b_gen, DoubleGen):
+        query_parts.append('SUM(b) OVER (ORDER BY a RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS sum_col')
+
+    if spark_version() > "3.1.1":
+        query_parts.append('NTH_VALUE(b, 1) IGNORE NULLS OVER '
+                           '(ORDER BY a RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS nth_1_ignore_nulls')
+
+    assert_gpu_and_cpu_are_equal_sql(
+        lambda spark : two_col_df(spark, UniqueLongGen(), b_gen, length=1024 * 14),
```
Comment (revans2) [inline integration_tests/src/main/python/window_function_test.py:540]: Slight problem here. We have been using a UniqueLongGen to try and avoid issues with non-deterministic ordering. But we are not accurately testing the range query, especially for current row unless we do have duplicate values in it.

**Review by revans2 — COMMENTED** [review]
In general it looks really good. Just want to see if there is something we can do with the testing for range windows.  Probably is fine as is, but I would love it if we could get some coverage.

File: integration_tests/src/main/python/window_function_test.py (hunk 2 above)
Comment (mythrocks) [reply integration_tests/src/main/python/window_function_test.py:541]: That's a good point.  The problem then becomes one of allowing repeated values in `ORDER BY`, without also introducing non-determinism.

One option is to use a non-`UniqueLongGen()` for the `ORDER BY` column, and using the same column as the aggregation column. This should ensure deterministic results, in the event of repeats.

Let me try this out, and update the PR.

P.S. The last commit combines the `ROW` and `RANGE` based running window tests.  Please disregard it.  I'd better keep these tests separate.

**Comment by revans2** [comment]
build

**Review by revans2 — APPROVED** [review]
Looks good to me.

**Comment by mythrocks** [comment]
Thanks for the review, @revans2.  I'm adding a couple of similar new tests for the partitioned cases.

File: integration_tests/src/main/python/window_function_test.py (hunk 2 above)
Comment (mythrocks) [reply integration_tests/src/main/python/window_function_test.py]: I've added tests with repeated values in the order-by columns.

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
Build

**Review by revans2 — APPROVED** [review]

**Pull request merged**